docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,220 @@
+Backward mode AD is then computing, for a given vector :math:`v` of size :math:`M`, the quantity :math:`v^T J_f`.
+Forward mode AD on the other hand is computing, for a given vector :math:`u` of size :math:`N`, the quantity :math:`J_f u`.
+
+For functions that contain complex values, the story is a lot more complex. We only provide the gist here and the full description can be found at :ref:`complex_autograd-doc`.
+
+Proper complex derivatives are too restrictive for most functions we care about. So we turn to Wirtinger calculus.
+In a basic setting of Wirtinger calculus, the chain rule required access to both the Wirtinger derivative (called :math:`W` below) and the Conjugate Wirtinger derivative (called :math:`CW` below). This means that in the general case of Wirtinger calculus, the values that we would need to "backward through the graph" in backward mode AD would need to be twice the size of the "forward value" (both :math:`W` and :math:`CW` need to be propagated because in general, despite their name, one is not the complex conjugate of the other).
+
+To avoid this problem, for backward mode AD, we always work under the assumption that the current function is part of a bigger function whose output is in :math:`\mathcal{R}`. Note that this assumption means that all the intermediary gradients we compute during the backward pass are also corresponding to functions whose output is in :math:`\mathcal{R}`.
+In practice, this assumption is not restrictive when doing optimization as such problem require a real-valued objective (as there is no natural ordering of the complex numbers).
+
+Under this assumption, we can show that :math:`W = CW^*` (we use :math:`*` to denote complex conjugation here) and so only one of the two values actually need to be "backwarded through the graph" as the other one can easily be recovered.
+To simplify internal computations, PyTorch uses :math:`2 * CW` as the value it backwards and returns when the user asks for gradients.
+Similarly to the real case, when the output is actually in :math:`\mathcal{R}^M`, backward mode AD does not compute :math:`2 * CW` but only :math:`v^T (2 * CW)` for a given vector :math:`v \in \mathcal{R}^M`.
+
+For forward mode AD, we use a similar logic, in this case, assuming that the function is part of a larger function whose input is in :math:`\mathcal{R}`. Under this assumption, we can make similar claim that every intermediary result corresponds to a function whose input is in :math:`\mathcal{R}` and in this case, we can show that :math:`W = CW` for the intermediary functions.
+To make sure the forward and backward mode compute the same quantities, the forward mode also computes :math:`2 * CW`.
+Similarly to the real case, when the input is actually in :math:`\mathcal{R}^N`, forward mode AD does not compute :math:`2 * CW` but only :math:`(2 * CW) u` for a given vector :math:`u \in \mathcal{R}^N`.
```

Comment:
```suggestion
```

Reply:
The above text didn't address what happens when the input is not in :math:`\mathcal{R}^N`. The answer is we assume it's a part of a real-valued function. But anyway, I think it's a repeat of things you have already stated and explained, and so can be skipped.

Reply:
It is very similar to the paragraph above but I think there are key differences in the fact that now the product is on the right and the space where `u` lives is not the same.

Also because we re-use these products below, I prefer to have them stated here so that people can very easily know where they come from.

Reply:
ahh okay